---
title: "Base R template for 16S amplicon processing"
output: 
  html_document: 
    keep_md: yes
    toc: yes
    toc_float: yes
    toc_collapsed: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
# Load required libraries
library(dada2)
library(Biostrings)
library(DECIPHER)
library(phyloseq)
library(tidyverse)
library(telegramNotify) # custom notification package

# Load helper functions
source('functions.R')
set.seed(5678)
```

# Read metadata

`read_metadata(filename, sample.names, ...)`

Read metadata, assign sample names as rownames (important for downstream analysis).

```{r}
mdat <- read_metadata(filename = "16s_map.csv",
                      sample.names = "Filename",
                      sep = ",",
                      check_files = "16s_raw/")
mdat
```

# ASV picking

Run DADA2 pipeline for 16S sequences.

```{r}
notified_reads_to_seqtable_16s <- notify_on_error(fun = reads_to_seqtable_16s, 
                                          task_name = "16S ASV picking",
                                          notify_success = TRUE)

seqtab <- notified_reads_to_seqtable_16s(raw_files_path = "16s_raw",
                                trimLeft = c(20, 19),
                                truncLen = c(240, 230), 
                                log_file = "16S_processing.log",
                                pool = "pseudo",
                                cores = TRUE,
                                maxEE = c(2,5),
                                pattern = c("_L2_1.fq.gz", "_L2_2.fq.gz"),
                                plot_profile = TRUE)

# telegramNotify::send_telegram_file("16S_processing.txt")
```

# Taxonomic assignment

Assign taxonomy using SILVA database.

```{r}
notified_assign_taxonomy <- notify_on_error(fun = assign_taxonomy, 
                                          task_name = "16S Taxonomy",
                                          notify_success = TRUE)

taxa <- notified_assign_taxonomy(seqtab = seqtab,
                        set_train_path = "/home/alexey/tax_n_refs/silva_nr_v138_train_set.fa.gz",
                        cores = TRUE)
taxa[1:10]
```

# Phyloseq object assembly

Combine ASV table, taxonomy, and metadata into a `phyloseq` object. Optionally, filter out organelle sequences and export reference FASTA.

```{r}
ps <- assemble_phyloseq_16s(seqtab = seqtab,
                            metadata = mdat, 
                            taxonomy = taxa,
                            filter.organelles = TRUE, 
                            write_fasta = 'refseqs.fasta')

ps
```

# Phylogenetic tree construction

Build tree from representative sequences and merge into `phyloseq` object.

```{r}
system2(c('./build_tree.sh', 'refseqs.fasta'))
tree <- ape::read.tree('tree.nwk')
ps <- merge_phyloseq(ps, tree)

saveRDS(ps, "bact.RData")
```

# Basic stats

Explore the dataset: number of taxa, reads per sample, taxonomy and ASV tables. Save `phyloseq` object to `.RData` file.


```{r}
sample_names(ps) # Names of samples
sample_sums(ps)  # Number of reads per sample

tax_table(ps)[1:5, 1:4] # First rows of taxonomy table
otu_table(ps)[1:4, 1:5] # First rows of ASV table

ps@sam_data # Metadata
```
