---
title: "Base R template for 16s amplicon processing"
output: html_notebook
---

```{r}
library(dada2)
library(Biostrings)
library(DECIPHER)
library(phyloseq)
library(ggplot2)
library(ggpubr)
library(ape)
library(dplyr)
library(DESeq2)

set.seed(5678)
setwd('/path/')
```

## Export phyloseq object and tuning

```{r}
ps <- readRDS()

sample_names(ps) # Names of samples
sample_sums(ps) # Sum of reads per sample


tax_table(ps)[1:5, 1:4] # Taxonomy table
otu_table(ps)[1:4, 1:5] # ASV table
```

```{r}
# Write table as .csv
write_ASVs_table <- function(ps_object, filename){
  write.csv(cbind(ps_object@otu_table %>% t() %>%  data.frame(),
                  ps_object@tax_table %>% data.frame()),
            filename)
}
```


## EDA

```{r}
# Draw barplot of relative abundance by taxa level
bargraph <- function(ps, rank, threshold=0.05){
  require(dplyr)
  require(ggplot2)
  require(phyloseq)
  
  ps <- prune_taxa(taxa_sums(ps) > 0, ps)
  ps2 <- tax_glom(ps, taxrank = rank)
  ps3 = transform_sample_counts(ps2, function(x) x / sum(x) )
  data <- psmelt(ps3) # create dataframe from phyloseq object
  data$Plot <- as.character(data[,rank]) # convert to character
  data$Plot[data$Abundance < threshold] <- paste0("<", threshold, " abund.")
  medians <- data %>% group_by(Plot) %>% mutate(median=median(data$Abundance))
  remainder <- medians[medians$median <= threshold,]$Plot
  
  # create palette long enough for our data
  base.palette <- c("darkblue", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "lightskyblue", 
               "darkgreen", "deeppink", "khaki2", "firebrick", "brown1", "darkorange1", "cyan1", "royalblue4", 
               "darksalmon", "dodgerblue3", "steelblue1", "darkgoldenrod1", "brown1", "cyan1", "darkgrey")
  required.colors <- nlevels(factor(data$Plot))
  repeats = required.colors %/% length(base.palette) + 1
  palette <- rep(base.palette, length.out = repeats * length(base.palette))
  
  p <- ggplot(data=data, aes(x=Sample, y=Abundance, fill=Plot))
  p + geom_bar(aes(), stat="identity", position="stack") + theme_light() +
    scale_fill_manual(values = palette) +
    theme(legend.position="bottom") + guides() +
    theme(axis.text.x = element_text(angle = 90))
}
```

```{r}
# Calculate several alpha-diversity indexes, return one dataframe
alpha_div_table <- function(ps, metric, group){
  require(phyloseq)
  
  ps <- prune_taxa(taxa_sums(ps) > 0, ps)
  
  obs_sim <- estimate_richness(ps, split = TRUE, measures = metric)
  alpha <- cbind(sample_data(ps)[,group], obs_sim)
  return(alpha)
}

# Plot alpha-diversity by selected metric
plot_alpha <- function(ps, group, metric) {
  require(phyloseq)
  require(ggplot2)
  
  ps_a <- prune_taxa(taxa_sums(ps) > 0, ps)
  plot_richness(ps_a, x=group, measures=metric) + 
    geom_boxplot() +
    geom_point(size=1.2, alpha=0.3) + 
    theme_light() + scale_color_brewer(palette="Dark2") +
    theme(axis.text.x = element_text(angle = 90),
          axis.title.x=element_blank()) +
    labs(y=paste(metric, "index")) +
    ggtitle(metric)
}
```

```{r}
# Plot beta-diversity
beta_plot <- function(ps, method, distance, ...){
  require(phyloseq)
  require(ggplot2)
  
  ps.prop <- transform_sample_counts(ps, function(x) x/sum(x))
  ords <- ordinate(ps.prop, method=method, distance=distance)
  plot_ordination(ps.prop, ords, title=deparse(substitute(ps)), ...) +
    geom_point(size=3, alpha=0.7) + 
    theme_light()
}
```

## DeSEQ

```{r}
# Gets table of significant differs from ps and formula
sig_table <- function(ps, formula, threshold){
  require(DESeq2)
  require(dplyr)
  
  ds <- phyloseq_to_deseq2(ps, formula)
  ds = estimateSizeFactors(ds, type="poscounts")
  ds = estimateDispersions(ds, fitType = "local")
  ds = DESeq(ds)
  #mcols(ds, use.names=TRUE)
  res = results(ds, cooksCutoff = FALSE)
  alpha = 0.05
  sigtab = res[which(res$padj < alpha), ]
  sigtab <- sigtab %>% data.frame() %>% dplyr::filter(baseMean >= threshold[1],
                                                      log2FoldChange >= threshold[2] | log2FoldChange <= threshold[2])
  if (nrow(sigtab) == 0) {
    return(NA)
  }
  sigtab = cbind(as(sigtab, "data.frame"),
                 as(tax_table(ps)[rownames(sigtab), ], "matrix")
  )
  return(sigtab)
}
```

```{r}
# Draw a plot by %sig_table% data
draw_sig_table <- function(sig_table, rank){
  require(ggplot2)
  
  sig_table$Class <-  ifelse(is.na(sig_table$Class), 
                             paste(sig_table$Phylum, "// NA"), 
                             paste(sig_table$Class))
  sig_table$Order <-  ifelse(is.na(sig_table$Order), 
                             paste(sig_table$Class, "// NA"), 
                             paste(sig_table$Order))
  sig_table$Family <- ifelse(is.na(sig_table$Family), 
                             paste(sig_table$Order, "// NA"), 
                             paste(sig_table$Family))
  sig_table$Genus <- ifelse(is.na(sig_table$Genus), 
                            paste(sig_table$Class, "// NA"), 
                            paste(sig_table$Genus))
  
  sig_table[sig_table == "Burkholderia-Caballeronia-Paraburkholderia"] <- "Burkholderia"
  sig_table[sig_table == "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium"] <- "Pararhizobium"
  
  ggplot(sig_table, aes(y = sig_table[,rank], x=log2FoldChange, size = log(baseMean))) + 
    geom_point(aes(color=sig_table[,rank])) + guides(colour=FALSE) +
    theme_light() + facet_grid(rows = vars(Phylum), space = 'free_y', scales = 'free') +
    theme(strip.text.y = element_text(angle=0)) +
    theme(legend.position = "top") + ylab(rank)
}
```

```{r}
# Plot heatmap on selected
plot_heatmap <- function(ps, group = "SampleID", log.transform = TRUE){
  require(dplyr)
  require(phyloseq)
  require(ggplot2)
  
  sig.taxa.long <- psmelt(sig.ps) %>%
    arrange(Phylum) %>% 
    mutate(row = row_number()) %>% 
    mutate_at(c('Genus', 'Family', 'Order', 'Class', 'Phylum'), as.character)  # fuckin magic for ifelse tier
  
  sig.taxa.long$Abundance <- as.numeric(sig.taxa.long$Abundance)
  sig.taxa.long$Taxa <- ifelse(is.na(sig.taxa.long$Genus),
                               ifelse(is.na(sig.taxa.long$Family), 
                                      ifelse(is.na(sig.taxa.long$Order), 
                                             ifelse(is.na(sig.taxa.long$Class), 
                                                    paste(sig.taxa.long$Phylum, "// NA"), 
                                                    paste(sig.taxa.long$Class, "// NA")),
                                             paste(sig.taxa.long$Order, "// NA")),
                                      paste(sig.taxa.long$Family, "// NA")), 
                               sig.taxa.long$Genus)
  sig.taxa.long[sig.taxa.long == "Burkholderia-Caballeronia-Paraburkholderia"] <- "Burkholderia"
  sig.taxa.long[sig.taxa.long == "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium"] <- "Pararhizobium"
  
  
  ggplot(sig.taxa.long, aes(x = sig.taxa.long[,group], y = reorder(Taxa, row))) + 
    {if(log.transform)geom_tile(aes(fill=log(Abundance)))} +
    {if(!log.transform)geom_tile(aes(fill=Abundance))} +
    scale_fill_distiller(palette = "OrRd", trans = "reverse") +
    facet_grid(rows = vars(Phylum), scales = "free_y", space = "free") +
    theme(strip.text.y = element_text(angle = 0),
          panel.spacing = unit(0.05,'lines')) +
    xlab("") + ylab("") +
    theme(axis.text.x = element_text(angle = 90))
}
```
